pkg/components/ebpf/common/mongo_detect_transform.go

```diff
@@ -0,0 +1,530 @@
+	}
+	pendingRequest, moreToCome, err = parseMongoMessage(buf, *header, time, isRequest, inFlightRequest)
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to parse MongoDB response: %w", err)
+	}
+	if pendingRequest == nil {
+		return nil, false, errors.New("no MongoDB request or response found in the message")
+	}
+	requests.Add(key, pendingRequest)
+	if !moreToCome && !isRequest {
+		requests.Remove(key)
+		// If this is a response and there are no more sections to come, we can finalize the request
+		return pendingRequest, false, nil
+	}
+	return nil, true, nil
+}
+
+func parseMongoMessage(buf []uint8, hdr msgHeader, time int64, isRequest bool, pendingRequest *MongoRequestValue) (*MongoRequestValue, bool, error) {
+	switch hdr.OpCode {
+	case opMsg:
+		return parseOpMessage(buf, time, isRequest, pendingRequest)
+	default:
+		return nil, false, fmt.Errorf("unsupported MongoDB operation code %d", hdr.OpCode)
+	}
+}
+
+func validateOpMsg(isRequest bool, flagBits int32, pendingRequest *MongoRequestValue) (bool, error) {
+	// TODO (mongo): maybe add checksum validation to avoid false positives? (only if we have the full packet)
+	moreToCome := flagBits&flagMoreToCome != 0
+	if isRequest {
+		switch {
+		case pendingRequest == nil:
+			return true, nil
+		case !pendingRequest.inRequest():
+			return false, errors.New("MongoDB request received already started receiving response")
+		case pendingRequest.Flags&flagMoreToCome == 0:
+			return false, errors.New("MongoDB request with moreToCome flag not set but got another request")
+		}
+	} else {
+		exhaustAllowed := pendingRequest.Flags&flagExhaustAllowed != 0
+		if moreToCome && !exhaustAllowed {
+			return false, errors.New("MongoDB response with moreToCome flag set but exhaustAllowed is not set")
+		}
+		if pendingRequest.inRequest() && pendingRequest.Flags&flagMoreToCome != 0 {
+			return false, errors.New("MongoDB request expects more sections but response is sent")
+		}
+	}
+	return moreToCome, nil
+}
+
+func addSectionToMessage(isRequest bool, pendingRequest *MongoRequestValue, sections []mongoSection, flagBits int32, time int64) (*MongoRequestValue, error) {
+	if isRequest {
+		if pendingRequest == nil {
+			pendingRequest = &MongoRequestValue{
+				RequestSections: sections,
+				StartTime:       time,
+				Flags:           flagBits,
+			}
+		} else {
+			pendingRequest.RequestSections = append(pendingRequest.RequestSections, sections...)
+			pendingRequest.Flags = flagBits
+			if pendingRequest.StartTime > time {
+				pendingRequest.StartTime = time
+			} else if pendingRequest.EndTime < time {
+				pendingRequest.EndTime = time
+			}
+		}
+	} else {
+		if pendingRequest == nil {
+			return nil, errors.New("MongoDB response received but no pending request found")
+		}
+		if pendingRequest.ResponseSections != nil {
+			pendingRequest.ResponseSections = append(pendingRequest.ResponseSections, sections...)
+		} else {
+			pendingRequest.ResponseSections = sections
+		}
```

Comment:
I think go allows writing
```suggestion
		pendingRequest.ResponseSections = append(pendingRequest.ResponseSections, sections...)
```
i.e. it deals with the nil case already